Algorithm paper (#1249)
pendulum-project/ntpd-rs: docs/algorithm/algorithm.tex

\documentclass{article}

\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{graphicx}

\DeclareMathOperator{\erf}{erf}
\DeclareMathOperator{\Var}{Var}
\DeclareMathOperator{\Cov}{Cov}

\begin{document}

\section{Global architecture}

The clock control filter of NTPD-rs consists of 4 stages.

> First, on a per-time-source basis, a kalman filter is used to estimate the offset and frequency error as compared to that specific timesource.

Reviewer: Nice intro, but this could use a graphic like the one on the whiteboard.

Author: Added graphic

During this stage, the remote clock is treated as ground truth and any uncertainty with respect to UTC it reports is ignored.
Sections~\ref{sec:kalmanfilter}-\ref{sec:additionalchecks} describe this kalman filter.

Next, the error the remote clock estimates itself to have compared to UTC is added to the error estimates.
This allows us to compare all time sources and see which subset of them is in agreement.
Selection criteria, as well as the used algorithm, are detailed in Section~\ref{sec:selection}.

Having selected which sources to use, these sources' measurements are averaged to produce an estimate of the clock offset.
The procedure for the averaging is described in Section~\ref{sec:averaging}.

Finally this clock offset, and the error bounds on it, are used to steer the systemclock to bring it in closer agreement with the remote time sources.
We steer primarily using frequency adjustment, the exact procedure used is described in Section \ref{sec:steering}.

\section{Kalman filter}\label{sec:kalmanfilter}

> To estimate offset and frequency offset with regard to a remote clock, we use a kalman filter to interpret the measurements.

Author: fixed

> To estimate offset and frequency offset with regard to a remote clock, we use a kalman filter to interpret the measurements.

Author: Added some text in the introduction that should fix this.

Although we regard the remote clock as ground truth for this filter, we choose to parameterize the state as a function of local time $t_l$.
We will use $t_r$ to denote the time indicated by the remote clock.
The remote time corresponding to local time $t_l$ is denoted as $t_r(t_l)$.
For convenience we will use the same symbols as Wikipedia for the equations for a kalman filter given below.

Our state model consists of two parameters: The offset $\Delta$ and the frequency difference $\omega$ to the remote clock. These are defined such that
\begin{align}
t_r(t_l) &= t_l + \Delta(t_l)\\

> \omega &= \frac{t_r(t_l') - t_r(t_l)}{t_l' - t_l} - 1

Author: changed

\end{align}
Combined, these form our state vector
\begin{align}
x &= \begin{pmatrix}
\Delta\\
\omega
\end{pmatrix}
\end{align}

> We will use the symbol $P$ to refer to the covariance matrix for this state.

Author: I think it is best to introduce it here to keep the flow better when starting to discuss the steps.

Given these definitions, the update matrix for when a local timespan $\delta$ has passed is given by
\begin{align}
F &= \begin{pmatrix}
1 & \delta\\
0 & 1
\end{pmatrix}
\end{align}
Updating the filters state from $t_l$ to $t_l+\delta$ is then given by
\begin{align}
x(t_l+\delta) &= Fx(t_l)\\
P(t_l+\delta) &= FPF^\mathrm{T} + Q
\end{align}
Here we will keep the process noise matrix $Q(\delta)$ general, stating only that we require two update steps of sizes $\delta_1$ and $\delta_2$
to give the same result as a single update step of size $\delta_1 + \delta_2$ when no measurement is done in between.
A detailed treatment of our noise model is given in Section~TODO.

Measurements from the NTP protocol give us an immediate estimate of $\Delta$. Letting $z$ be the (1-element) vector with this measurement,
and $R$ it's covariance matrix (which in this case is simply an estimate of the variance of the measurement).
This corresponds to a measurment matrix of the form
\begin{align}
H &= \begin{pmatrix}
1 & 0
\end{pmatrix}
\end{align}
Such a measurment can be incorporated into the filter by first calculating
\begin{align}
y &= z - Hx\\
S &= H P H^\mathrm{T} + R\\

> K &= P H S^{-1}

Author: Added description below what these matrices represent.

\end{align}
which allows us to update the state through
\begin{align}
x' &= x + Ky\\
P' &= (I - K H)P
\end{align}
where $x'$ and $P'$ denote the post-measurement state and covariance matrix, and $I$ is the identity matrix.

\section{Process noise}\label{sec:processnoise}

For our kalman filter we need estimates for the process noise $Q$ and measurement noise $R$. Let us start by first considering the process noise.

Working out the requirement that a single large step should give the same as two smaller steps, we find that for all $x$, (symmetric) $P$, $\delta_1$ and $\delta_2$
we should have
\begin{align}
F(\delta_1 + \delta_2)x = F(\delta_2)F(\delta_1)x\\
F(\delta_1 + \delta_2)PF(\delta_1 + \delta_2)^\mathrm{T} + Q(\delta_1+\delta_2) =\nonumber\\

> F(\delta_2)(F(\delta_1)PF(\delta_1)^\mathrm{T} + Q(\delta_1))F(\delta_2)^\mathrm{T} + Q(\delta_2)

Author: Refers to the same F as defined above. Added additional text stating that we keep the names around from the kalman section.

\end{align}

The first follows trivially from the identity
\begin{align}
F(\delta_1 + \delta_2) = F(\delta_2)F(\delta_1)
\end{align}

This same identity can also be used to simplify the second requirment to
\begin{align}
Q(\delta_1+\delta_2) = F(\delta_2)Q(\delta_1)F(\delta_2)^\mathrm{T}+Q(\delta_2).
\end{align}
Writing out this equation in the components $q_{ij}(\delta_t)$ (using that $Q$ is symmetric) gives
\begin{align}
q_{11}(\delta_1+\delta_2) &= q_{11}(\delta_1) + q_{11}(\delta_2) + 2\delta_2q_{12}(\delta_1)+\delta_2^2q_{22}(\delta_1)\\
q_{12}(\delta_1+\delta_2) &= q_{12}(\delta_1) + q_{12}(\delta_2) + \delta_2q_{22}(\delta_1)\\
q_{22}(\delta_1+\delta_2) &= q_{22}(\delta_1) + q_{22}(\delta_2)
\end{align}
Solving these equations, we find that there need to be constants $A$, $B$ and $C$ such that 
\begin{align}
q_{11}(\delta) &= \frac{A}{3}\delta^3 + B\delta^2 + C\delta\\
q_{12}(\delta) &= \frac{A}{2}\delta^2 + B\delta\\
q_{22}(\delta) &= A\delta
\end{align}

The $A$ contribution to $Q$ corresponds with a random walk noise of the frequency with infinitesimal steps.

> Similarly, the $C$ contribution is a random walk noise of the phase, without any underlying frequency change.

Author: Added appendix explaining this.

estimate it's own process noise.

Our assumption is reasonable for NTP as for modern computer oscillators, the phase random walk noise is dwarfed by channel noise until frequency random
walk becomes dominant. 
We assume the $B$ component to be irrelevant based on the observations in the previous section.

\subsection{Estimating process noise}

The process noise for our kalman filter will depend both on characteristics of the remote clock as well as those of the local oscillator.
As such, although we can estimate a reasonable value as a starting point, it is unlikely that all computers will perform at their best with
a single estimate for the frequency random walk noise.

To solve this, we dynamically determine the process noise at runtime.
The filter is started with an initial estimate of the wander noise, by default $A=10^{-16}$.
Then, as each measurement comes in, we calculate the probability $p$ of getting a measurement that close or closer to our state estimate.
Based on the distribution of $p$ we can then determine whether our estimate of the noise is low (if $p$ consistently is large) or high (if $p$ consistently is small) and adjust $A$ accordingly.

> Concretely, we keep a counter $C$ with a starting value of $0$. If $p$ is smaller than $\frac{1}{3}$, we substract one from $C$.

Author: Fixed

Conversely, if $p$ is larger than $\frac{2}{3}$, we add one to $C$. Finally, if $p$ is between $\frac{1}{3}$ and $\frac{2}{3}$, we move $C$ one towards $0$.

If $C$ gets above $16$, we reset $C$ to $0$ and multiply $A$ by four (doubling the standard deviation of the noise).
Similarly, if $C$ gets below $-16$, we divide $A$ by four.
Waiting for multiple measurements before adjusting our uncertainty estimates ensures that our estimate doesn't keep changing.
This is further helped by having the region between $\frac{1}{3}$ and $\frac{2}{3}$ push $C$ towards $0$, as this works to keep $|C|$ small.

Of course, $p$ shows problems with the process noise only when it is a significant fraction of the total measurement noise.
As such, if the measurement noise is more than $\frac{9}{10}$ the total noise of the difference between measurement and prediction, a $p$ smaller
than $\frac{1}{3}$ no longer lowers $C$, instead also pushing it towards $|C|$, to avoid becoming overly optimistic on process noise if the
measurement noise estimate is somewhat pessimistic.

\subsubsection{Calculation details for $p$}

> To find $p$, first note that $S$ is effectively the covariance matrix of the difference $y$ between the prediction from the filter for the measurement and

Author: Added  explanation and note that we keep symbol names.

> the actual measured value. This implies that $y S y^\mathrm{T}$ is distributed according to a chi-squared distribution with one degree of freedom.

Author: added notes that we keep notation from kalman filter section.

\documentclass{article}

\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{graphicx}
\usepackage{tikz}

\DeclareMathOperator{\erf}{erf}
\DeclareMathOperator{\Var}{Var}
\DeclareMathOperator{\Cov}{Cov}
\DeclareMathOperator{\E}{E}

\title{Clock steering in ntpd-rs}

\begin{document}

\maketitle

> In NTP, each message sent to an NTP server gives a single response. For this roundtrip, the server reports back at what time it received the request and when it sent the response. These timestamps, together with the when the client actually sent the request and response, together give a measure of the offset between the clock of the client and the server, and of how long the network delay between them was.

Author: fixed

	\draw (1,3) rectangle (3.5, 4) node[pos = .5] {kalman filter};
	\draw (1,4.5) rectangle (3.5, 5.5) node[pos = .5] {kalman filter};
	\draw[->] (3.5, 0.5) -- (4.5, 0.5);
	\draw[->] (3.5, 2) -- (4.5, 2);
	\draw[->] (3.5, 3.5) -- (4.5, 3.5);
	\draw[->] (3.5, 5) -- (4.5, 5);
	\draw (4.5, 0) rectangle (5.5, 5.5) node[pos = .5, rotate=-90] {selection};
	\draw[->] (5.5, 0.5) -- (6.5, 0.5);
	\draw[->] (5.5, 3.5) -- (6.5, 3.5);
	\draw[->] (5.5, 5) -- (6.5, 5);
	\draw (6.5, 0) rectangle (7.5, 5.5) node[pos = .5, rotate=-90] {combine};
	\draw[->] (7.5, 2.75) -- (8.5, 2.75);
	\draw (8.5, 2) rectangle (11.5, 3.5) node[pos = .5] {steer};
	\end{tikzpicture}
}
\caption{Diagram with the parts of the ntpd-rs clock control filter. The arrows on the left represent the measurements from the (in this case 4) time sources coming in.}
\end{figure}

The clock control filter of ntpd-rs consists of 4 stages.

First, on a per-time-source basis, a kalman filter is used to estimate the offset and frequency error as compared to that specific timesource.
During this stage, the remote clock is treated as ground truth and any uncertainty with respect to UTC it reports is ignored.

> Sections~\ref{sec:kalmanfilter}-\ref{sec:additionalchecks} describe this kalman filter.

Author: The extra s on describes shouldn't be there, applied the Kalman suggestion.

> Next, the error the remote clock estimates itself to have compared to UTC is added to the error estimates.

Author: Placed in the correct location and rephrased a bit more, needs one more review pass.

Then the overal desired poll interval is the minimum of the desired poll intervals of all the time sources. This is currently done as a global indicator primarily because of limitations in the architecture of ntpd-rs. In the future, this may change to a desired poll interval that is different for each peer.

Note also that, as in the case of process noise estimation, the specific numbers mentioned above are configurable. The values above are the current default values used by ntpd-rs.

\section{Performance}

We evaluated the performance of the new algorithm against the default settings of chrony version 4.0, another NTP client. For this, we connected a raspberry pi 4 via a switch to a local gps-connected ntp server (an Endrun Ninja). Both the raspberry pi and the ntp server were configured to produce pulse per second outputs.

> For each test, the NTP client under test was started on the raspberry pi with a fixed poll interval of 1 second and allowed to stabilize over a period of at least half an hour. Then, the offset between the pulse per second signals of the raspberry pi were measured for one hour. The results of these measurements are shown in Figure~\ref{fig:offset-time}. Both chrony and ntpd-rs were configured to use the local server as only timesource, but otherwise used their default configuration.

Reviewer 2: ```suggestion
For each test, the NTP client under test was started on the Raspberry Pi with a fixed poll interval of 1 second and allowed to stabilize over a period of at least half an hour. Then, the offset between the pulse per second signals of the Raspberry Pi were measured for one hour. The results of these measurements are shown in Figure~\ref{fig:offset-time}. Both chrony and ntpd-rs were configured to use the local server as only timesource, but otherwise used their default configuration.
```

\begin{figure}
\includegraphics[width=0.5\textwidth]{offset-ntpd-rs.png}\includegraphics[width=0.5\textwidth]{offset-chrony.png}
\caption{Offset of the raspberry pi running ntpd-rs (left) and chrony (right) over a time period of 1 hour, compared to the server they were synchronizing to.}\label{fig:offset-time}
\end{figure}

> We haven't corrected in any way for assymetries in the setup on the network layer, so any difference between the average offset to the server and the raspberry pi cannot be interpreted as significant advantages for either ntpd-rs or chrony. However, the deviation from this mean does provide a clear indication of synchronization quality. For chrony, the deviation is $3.6$ microseconds. For ntpd-rs it is $1.8$ microseconds. From this we conclude that ntpd-rs is at least on par with chrony, and likely slightly better.

Reviewer 2: ```suggestion
We haven't corrected in any way for asymmetries in the setup on the network layer, so any difference between the average offset to the server and the Raspberry Pi cannot be interpreted as significant advantages for either ntpd-rs or chrony. However, the deviation from this mean does provide a clear indication of synchronization quality. For chrony, the deviation is $3.6$ microseconds. For ntpd-rs it is $1.8$ microseconds. From this we conclude that ntpd-rs is at least on par with chrony, and likely slightly better.
```

	\draw[->] (3.5, 2) -- (4.5, 2);
	\draw[->] (3.5, 3.5) -- (4.5, 3.5);
	\draw[->] (3.5, 5) -- (4.5, 5);
	\draw (4.5, 0) rectangle (5.5, 5.5) node[pos = .5, rotate=-90] {selection};
	\draw[->] (5.5, 0.5) -- (6.5, 0.5);
	\draw[->] (5.5, 3.5) -- (6.5, 3.5);
	\draw[->] (5.5, 5) -- (6.5, 5);
	\draw (6.5, 0) rectangle (7.5, 5.5) node[pos = .5, rotate=-90] {combine};
	\draw[->] (7.5, 2.75) -- (8.5, 2.75);
	\draw (8.5, 2) rectangle (11.5, 3.5) node[pos = .5] {steer};
	\end{tikzpicture}
}
\caption{Diagram with the parts of the ntpd-rs clock control filter. The arrows on the left represent the measurements from the (in this case) 4 time sources coming in.}
\end{figure}

The clock control filter of ntpd-rs consists of 4 stages.

First, on a per-timesource basis, a Kalman filter is used to estimate the offset and frequency error as compared to that specific timesource.
During this stage, the remote clock is treated as ground truth and any uncertainty with respect to UTC it reports is ignored.
Sections~\ref{sec:Kalmanfilter}-\ref{sec:additionalchecks} describe this Kalman filter.

Next, we compare all time sources and see which subset of them is in agreement.

> We also check for each time source whether it is of sufficient quality to us.e

Reviewer 2: ```suggestion
We also check for each time source whether it is of sufficient quality to use.
```

Selection criteria, as well as the used algorithm, are detailed in Section~\ref{sec:selection}.

Having selected which sources to use, these sources' measurements are averaged to produce an estimate of the clock offset.

> A this stage, we also add the error compared to UTC the remote clock estimates itself to the error estimates.

Reviewer 2: I don't understand this sentence grammar wise

> A this stage, we also add the error compared to UTC the remote clock estimates itself to the error estimates.

Author: ```suggestion
```
